File: data/workflows/scrnaseq-scanpy-core-analysis/scripts/pseudobulk_de.py
"""
Pseudobulk Differential Expression Analysis

This module implements pseudobulk aggregation and differential expression analysis
for single-cell RNA-seq data using DESeq2.

For methodology and best practices, see references/pseudobulk_de_guide.md

Functions:
  - aggregate_to_pseudobulk(): Aggregate counts per sample × cell type
  - run_deseq2_analysis(): Run DESeq2 on pseudobulk data
  - export_de_results(): Export DE results to CSV
  - plot_volcano(): Create volcano plots
  - plot_ma(): Create MA plots
"""

import scanpy as sc
import numpy as np
import pandas as pd
from pathlib import Path
from typing import Dict, List, Optional, Union, Tuple
import warnings


def aggregate_to_pseudobulk(
    adata: sc.AnnData,
    sample_key: str,
    celltype_key: str,
    min_cells: int = 10,
    min_counts: int = 1,
    layer: Optional[str] = None
) -> Dict[str, pd.DataFrame]:
    """
    Aggregate single-cell counts to pseudobulk (sum per sample × cell type).

    CRITICAL: Use raw counts, not normalized. Use sum aggregation, not mean.

    Parameters
    ----------
    adata : AnnData
        AnnData object with raw counts
    sample_key : str
        Column in adata.obs with sample IDs
    celltype_key : str
        Column in adata.obs with cell type labels
    min_cells : int, optional
        Minimum cells per sample-celltype combination (default: 10)
    min_counts : int, optional
        Minimum total counts per sample-celltype (default: 1)
    layer : str, optional
        Layer to aggregate (default: None uses .X)

    Returns
    -------
    dict
        Dictionary with keys:
        - 'counts': DataFrame of aggregated counts (genes × samples)
        - 'metadata': DataFrame of sample metadata
        - 'n_cells': DataFrame of cell counts per sample-celltype

    Notes
    -----
    Filters out sample-celltype combinations with <min_cells cells.
    """
    print(f"Aggregating to pseudobulk...")
    print(f"  Sample column: {sample_key}")
    print(f"  Cell type column: {celltype_key}")

    # Check required columns
    if sample_key not in adata.obs.columns:
        raise ValueError(f"'{sample_key}' not found in adata.obs")
    if celltype_key not in adata.obs.columns:
        raise ValueError(f"'{celltype_key}' not found in adata.obs")

    # Get counts
    if layer is not None:
        counts = adata.layers[layer]
    else:
        counts = adata.X

    # Convert to dense if sparse
    if hasattr(counts, 'toarray'):
        counts = counts.toarray()

    # Create sample-celltype combinations
    adata.obs['sample_celltype'] = (
        adata.obs[sample_key].astype(str) + '_' +
        adata.obs[celltype_key].astype(str)
    )

    # Aggregate counts
    pseudobulk_counts = {}
    n_cells_dict = {}
    metadata_list = []

    for sample in adata.obs[sample_key].unique():
        for celltype in adata.obs[celltype_key].unique():
            # Select cells
            mask = (
                (adata.obs[sample_key] == sample) &
                (adata.obs[celltype_key] == celltype)
            )
            n_cells = mask.sum()

            # Filter by minimum cells
            if n_cells < min_cells:
                continue

            # Sum counts
            sample_celltype_id = f"{sample}_{celltype}"
            summed_counts = counts[mask, :].sum(axis=0)

            # Filter by minimum counts
            if summed_counts.sum() < min_counts:
                continue

            pseudobulk_counts[sample_celltype_id] = summed_counts
            n_cells_dict[sample_celltype_id] = n_cells

            # Store metadata
            metadata_list.append({
                'sample_celltype': sample_celltype_id,
                'sample': sample,
                'celltype': celltype,
                'n_cells': n_cells
            })

    # Convert to DataFrames
    counts_df = pd.DataFrame(pseudobulk_counts, index=adata.var_names)
    metadata_df = pd.DataFrame(metadata_list).set_index('sample_celltype')

    print(f"\nPseudobulk aggregation complete:")
    print(f"  Total sample-celltype combinations: {counts_df.shape[1]}")
    print(f"  Genes: {counts_df.shape[0]}")
    print(f"  Median cells per combination: {metadata_df['n_cells'].median():.0f}")

    return {
        'counts': counts_df,
        'metadata': metadata_df,
        'n_cells': metadata_df['n_cells']
    }


def run_deseq2_analysis(
    pseudobulk: Dict[str, pd.DataFrame],
    sample_metadata: pd.DataFrame,
    formula: str,
    contrast: List[str],
    celltype_key: str = 'celltype',
    output_dir: Union[str, Path] = "results/pseudobulk_de",
    use_rpy2: bool = True
) -> Dict[str, pd.DataFrame]:
    """
    Run DESeq2 differential expression analysis for each cell type.

    Parameters
    ----------
    pseudobulk : dict
        Output from aggregate_to_pseudobulk()
    sample_metadata : DataFrame
        Sample-level metadata with condition, batch, etc.
        Must have 'sample' column matching pseudobulk sample IDs
    formula : str
        DESeq2 design formula (e.g., "~ batch + condition")
    contrast : list of str
        DESeq2 contrast (e.g., ["condition", "treated", "control"])
    celltype_key : str, optional
        Column name for cell types (default: 'celltype')
    output_dir : str or Path
        Directory to save results
    use_rpy2 : bool, optional
        Use R DESeq2 via rpy2 (default: True)
        If False, uses pydeseq2 (pure Python, less tested)

    Returns
    -------
    dict
        Dictionary mapping cell type to DESeq2 results DataFrame

    Notes
    -----
    Requires R and DESeq2 installed if use_rpy2=True.
    Install in R: BiocManager::install("DESeq2")
    """
    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    # Get counts and metadata
    counts_df = pseudobulk['counts']
    pb_metadata = pseudobulk['metadata']

    # Merge with sample metadata
    pb_metadata = pb_metadata.merge(
        sample_metadata,
        left_on='sample',
        right_on='sample',
        how='left'
    )

    # Get unique cell types
    celltypes = pb_metadata[celltype_key].unique()
    print(f"\nRunning DESeq2 for {len(celltypes)} cell types...")

    de_results = {}

    for celltype in celltypes:
        print(f"\n  Cell type: {celltype}")

        # Subset to cell type
        celltype_mask = pb_metadata[celltype_key] == celltype
        celltype_counts = counts_df.loc[:, celltype_mask]
        celltype_metadata = pb_metadata[celltype_mask]

        n_samples = celltype_counts.shape[1]
        print(f"    Samples: {n_samples}")

        # Check minimum samples
        if n_samples < 3:
            print(f"    Skipping: <3 samples")
            continue

        # Run DESeq2
        if use_rpy2:
            results_df = _run_deseq2_rpy2(
                celltype_counts,
                celltype_metadata,
                formula,
                contrast
            )
        else:
            results_df = _run_deseq2_pydeseq2(
                celltype_counts,
                celltype_metadata,
                formula,
                contrast
            )

        if results_df is not None:
            de_results[celltype] = results_df

            n_sig = (results_df['padj'] < 0.05).sum()
            print(f"    Significant genes (padj<0.05): {n_sig}")

    return de_results


def _run_deseq2_rpy2(
    counts: pd.DataFrame,
    metadata: pd.DataFrame,
    formula: str,
    contrast: List[str]
) -> pd.DataFrame:
    """Run DESeq2 via rpy2."""
    try:
        import rpy2.robjects as ro
        from rpy2.robjects import pandas2ri
        from rpy2.robjects.packages import importr
    except ImportError:
        raise ImportError("rpy2 required. Install with: pip install rpy2")

    # Activate pandas conversion
    pandas2ri.activate()

    # Import R packages
    try:
        deseq2 = importr('DESeq2')
        base = importr('base')
    except Exception as e:
        raise ImportError(f"DESeq2 not found in R. Install with: BiocManager::install('DESeq2')\n{e}")

    # Convert to R objects
    r_counts = pandas2ri.py2rpy(counts.astype(int))
    r_metadata = pandas2ri.py2rpy(metadata)

    # Create DESeqDataSet
    ro.globalenv['counts_matrix'] = r_counts
    ro.globalenv['col_data'] = r_metadata

    ro.r(f'''
    dds <- DESeqDataSetFromMatrix(
        countData = counts_matrix,
        colData = col_data,
        design = {formula}
    )
    ''')

    # Filter low count genes
    ro.r('''
    keep <- rowSums(counts(dds) >= 10) >= 3
    dds <- dds[keep,]
    ''')

    # Run DESeq2
    ro.r('dds <- DESeq(dds, quiet=TRUE)')

    # Extract results
    contrast_str = f"c('{contrast[0]}', '{contrast[1]}', '{contrast[2]}')"
    ro.r(f'res <- results(dds, contrast={contrast_str})')

    # Shrink log2FoldChange
    ro.r(f'''
    res_shrunk <- lfcShrink(dds,
                           contrast={contrast_str},
                           res=res,
                           type="ashr",
                           quiet=TRUE)
    ''')

    # Convert to pandas
    results_df = pandas2ri.rpy2py(ro.r('as.data.frame(res_shrunk)'))
    results_df.index.name = 'gene'
    results_df = results_df.reset_index()

    pandas2ri.deactivate()

    return results_df


def _run_deseq2_pydeseq2(
    counts: pd.DataFrame,
    metadata: pd.DataFrame,
    formula: str,
    contrast: List[str]
) -> pd.DataFrame:
    """Run DESeq2 via pydeseq2 (pure Python)."""
    try:
        from pydeseq2.dds import DeseqDataSet
        from pydeseq2.ds import DeseqStats
    except ImportError:
        raise ImportError("pydeseq2 required. Install with: pip install pydeseq2")

    # Create DESeqDataSet
    dds = DeseqDataSet(
        counts=counts.T.astype(int),
        metadata=metadata,
        design_factors=formula.replace('~', '').strip().split('+')
    )

    # Run DESeq2
    dds.deseq2()

    # Compute statistics
    stat_res = DeseqStats(dds, contrast=contrast)
    stat_res.summary()

    results_df = stat_res.results_df
    results_df.index.name = 'gene'
    results_df = results_df.reset_index()

    return results_df


def export_de_results(
    de_results: Dict[str, pd.DataFrame],
    output_dir: Union[str, Path] = "results/pseudobulk_de",
    padj_threshold: float = 0.05,
    log2fc_threshold: float = 0
):
    """
    Export DE results to CSV files.

    Parameters
    ----------
    de_results : dict
        Dictionary mapping cell type to results DataFrame
    output_dir : str or Path
        Output directory
    padj_threshold : float, optional
        Adjusted p-value threshold (default: 0.05)
    log2fc_threshold : float, optional
        Absolute log2 fold change threshold (default: 0)
    """
    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    print(f"\nExporting DE results to {output_dir}")

    for celltype, results_df in de_results.items():
        # Export full results
        results_file = output_dir / f"{celltype}_deseq2_results.csv"
        results_df.to_csv(results_file, index=False)

        # Export significant genes
        sig_mask = (
            (results_df['padj'] < padj_threshold) &
            (results_df['log2FoldChange'].abs() > log2fc_threshold)
        )
        sig_df = results_df[sig_mask].sort_values('padj')

        sig_file = output_dir / f"{celltype}_deseq2_sig.csv"
        sig_df.to_csv(sig_file, index=False)

        print(f"  {celltype}: {len(sig_df)} significant genes")


def plot_volcano(
    results_df: pd.DataFrame,
    celltype: str,
    output_dir: Union[str, Path] = "results/pseudobulk_de",
    padj_threshold: float = 0.05,
    log2fc_threshold: float = 0.5,
    top_genes: int = 10
):
    """
    Create volcano plot for DE results.

    Parameters
    ----------
    results_df : DataFrame
        DESeq2 results
    celltype : str
        Cell type name
    output_dir : str or Path
        Output directory
    padj_threshold : float
        Adjusted p-value threshold for significance
    log2fc_threshold : float
        Log2 fold change threshold for significance
    top_genes : int
        Number of top genes to label
    """
    from plotnine import (ggplot, aes, geom_point, geom_hline, geom_vline,
                         labs, scale_color_manual, theme_minimal)
    from plotnine_prism import theme_prism
    try:
        from adjustText import adjust_text
        HAS_ADJUSTTEXT = True
    except ImportError:
        HAS_ADJUSTTEXT = False

    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    # Prepare data
    plot_df = results_df.copy()
    plot_df['-log10(padj)'] = -np.log10(plot_df['padj'])

    # Classify genes
    plot_df['significance'] = 'NS'
    sig_mask = (
        (plot_df['padj'] < padj_threshold) &
        (plot_df['log2FoldChange'].abs() > log2fc_threshold)
    )
    plot_df.loc[sig_mask, 'significance'] = 'Significant'

    # Get top genes
    top_df = plot_df.nsmallest(top_genes, 'padj')

    # Create plot
    p = (
        ggplot(plot_df, aes(x='log2FoldChange', y='-log10(padj)'))
        + geom_point(aes(color='significance'), alpha=0.5, size=1)
        + geom_hline(yintercept=-np.log10(padj_threshold), linetype='dashed', color='red')
        + geom_vline(xintercept=log2fc_threshold, linetype='dashed', color='red')
        + geom_vline(xintercept=-log2fc_threshold, linetype='dashed', color='red')
        + scale_color_manual(values={'NS': '#CCCCCC', 'Significant': '#E31A1C'})
        + labs(
            title=f'Volcano Plot: {celltype}',
            x='log2 Fold Change',
            y='-log10(adjusted p-value)'
        )
        + theme_prism()
    )

    # Add gene labels with adjustText if available
    if HAS_ADJUSTTEXT and len(top_df) > 0:
        # Draw base plot to get matplotlib figure
        import matplotlib.pyplot as plt
        fig = p.draw()
        ax = fig.axes[0]

        # Add labels with adjustText for non-overlapping placement
        texts = [
            ax.text(row['log2FoldChange'], row['-log10(padj)'],
                   row.name if hasattr(row, 'name') else str(i),
                   fontsize=8, alpha=0.9)
            for i, row in top_df.iterrows()
        ]
        adjust_text(
            texts,
            arrowprops=dict(arrowstyle='->', color='gray', lw=0.5, alpha=0.7),
            expand_points=(1.5, 1.5),
            force_text=(0.5, 0.5)
        )

        # Save the adjusted figure
        fig.savefig(output_dir / f"{celltype}_volcano.svg", dpi=300, bbox_inches='tight')
        fig.savefig(output_dir / f"{celltype}_volcano.png", dpi=300, bbox_inches='tight')
        plt.close(fig)
    else:
        # Save without labels if adjustText not available
        p.save(output_dir / f"{celltype}_volcano.svg", dpi=300, width=8, height=6)


def plot_ma(
    results_df: pd.DataFrame,
    celltype: str,
    output_dir: Union[str, Path] = "results/pseudobulk_de",
    padj_threshold: float = 0.05
):
    """
    Create MA plot for DE results.

    Parameters
    ----------
    results_df : DataFrame
        DESeq2 results
    celltype : str
        Cell type name
    output_dir : str or Path
        Output directory
    padj_threshold : float
        Adjusted p-value threshold for significance
    """
    from plotnine import (ggplot, aes, geom_point, geom_hline,
                         labs, scale_color_manual)
    from plotnine_prism import theme_prism

    output_dir = Path(output_dir)

    # Prepare data
    plot_df = results_df.copy()

    # Classify genes
    plot_df['significance'] = 'NS'
    plot_df.loc[plot_df['padj'] < padj_threshold, 'significance'] = 'Significant'

    # Create plot
    p = (
        ggplot(plot_df, aes(x='baseMean', y='log2FoldChange'))
        + geom_point(aes(color='significance'), alpha=0.5, size=1)
        + geom_hline(yintercept=0, linetype='dashed', color='black')
        + scale_color_manual(values={'NS': '#CCCCCC', 'Significant': '#E31A1C'})
        + labs(
            title=f'MA Plot: {celltype}',
            x='Mean Expression (log10)',
            y='log2 Fold Change'
        )
        + theme_prism()
    )

    # Save
    p.save(output_dir / f"{celltype}_ma.svg", dpi=300, width=8, height=6)


# Example usage
if __name__ == "__main__":
    # Example workflow
    print("Example pseudobulk DE workflow:")
    print("1. Aggregate counts: pseudobulk = aggregate_to_pseudobulk(adata, 'sample', 'celltype')")
    print("2. Run DESeq2: de_results = run_deseq2_analysis(pseudobulk, metadata, '~ condition', ['condition', 'treated', 'control'])")
    print("3. Export: export_de_results(de_results)")
    print("4. Plot: plot_volcano(de_results['CellType'], 'CellType')")

File: data/skills/scrnaseq-seurat-core-analysis/scripts/pseudobulk_de.R
#' Pseudobulk Differential Expression Analysis
#'
#' This module implements pseudobulk aggregation and differential expression
#' analysis for single-cell RNA-seq data using DESeq2 and muscat framework.
#'
#' For methodology and best practices, see references/pseudobulk_de_guide.md

library(Seurat)
library(DESeq2)
library(muscat)
library(SingleCellExperiment)
library(ggplot2)
library(ggprism)
library(dplyr)


#' Aggregate counts to pseudobulk (sum per sample × cell type)
#'
#' @param seurat_obj Seurat object with cell type annotations
#' @param sample_col Column in metadata with sample IDs
#' @param celltype_col Column in metadata with cell type labels
#' @param min_cells Minimum cells per sample-celltype (default: 10)
#' @param min_counts Minimum total counts per sample-celltype (default: 1)
#' @param assay Assay to use (default: "RNA")
#'
#' @return SingleCellExperiment object with pseudobulk data
#'
#' @export
aggregate_to_pseudobulk <- function(
  seurat_obj,
  sample_col = "sample_id",
  celltype_col = "cell_type",
  min_cells = 10,
  min_counts = 1,
  assay = "RNA"
) {
  cat("Aggregating to pseudobulk...\n")
  cat(sprintf("  Sample column: %s\n", sample_col))
  cat(sprintf("  Cell type column: %s\n", celltype_col))

  # Convert to SingleCellExperiment
  sce <- as.SingleCellExperiment(seurat_obj, assay = assay)

  # Add required columns
  colData(sce)$sample_id <- seurat_obj@meta.data[[sample_col]]
  colData(sce)$cluster_id <- seurat_obj@meta.data[[celltype_col]]

  # Aggregate using muscat (sum counts per sample × cluster)
  pb <- aggregateData(
    sce,
    assay = "counts",
    fun = "sum",  # CRITICAL: use sum, not mean
    by = c("cluster_id", "sample_id")
  )

  # Filter low-count combinations
  # This is done during DE analysis

  cat("\nPseudobulk aggregation complete:\n")
  cat(sprintf("  Cell types: %d\n", length(assayNames(pb))))
  cat(sprintf("  Samples: %d\n", ncol(pb)))

  return(pb)
}


#' Run DESeq2 differential expression for each cell type
#'
#' @param pseudobulk_data SingleCellExperiment from aggregate_to_pseudobulk()
#' @param formula DESeq2 design formula (e.g., "~ batch + condition")
#' @param contrast DESeq2 contrast vector (e.g., c("condition", "treated", "control"))
#' @param output_dir Directory to save results
#' @param min_samples Minimum samples per group (default: 3)
#' @param shrink_lfc Shrink log2 fold changes (default: TRUE)
#'
#' @return List of DESeq2 results, one per cell type
#'
#' @export
run_pseudobulk_deseq2 <- function(
  pseudobulk_data,
  formula = "~ condition",
  contrast = c("condition", "treated", "control"),
  output_dir = "results/pseudobulk_de",
  min_samples = 3,
  shrink_lfc = TRUE
) {
  dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)

  cat("\nRunning DESeq2 for each cell type...\n")

  # Get cell types (assays in muscat object)
  celltypes <- assayNames(pseudobulk_data)

  de_results <- list()

  for (celltype in celltypes) {
    cat(sprintf("\n  Cell type: %s\n", celltype))

    # Extract counts for this cell type
    counts <- assay(pseudobulk_data, celltype)
    metadata <- colData(pseudobulk_data)

    # Check sample size
    n_samples <- ncol(counts)
    cat(sprintf("    Samples: %d\n", n_samples))

    if (n_samples < min_samples) {
      cat(sprintf("    Skipping: <%d samples\n", min_samples))
      next
    }

    # Create DESeqDataSet
    dds <- tryCatch({
      DESeqDataSetFromMatrix(
        countData = counts,
        colData = metadata,
        design = as.formula(formula)
      )
    }, error = function(e) {
      cat(sprintf("    Error creating DESeqDataSet: %s\n", e$message))
      return(NULL)
    })

    if (is.null(dds)) next

    # Filter low count genes
    keep <- rowSums(counts(dds) >= 10) >= 3
    dds <- dds[keep, ]

    cat(sprintf("    Genes after filtering: %d\n", nrow(dds)))

    # Run DESeq2
    dds <- tryCatch({
      DESeq(dds, quiet = TRUE)
    }, error = function(e) {
      cat(sprintf("    Error running DESeq2: %s\n", e$message))
      return(NULL)
    })

    if (is.null(dds)) next

    # Extract results
    res <- results(dds, contrast = contrast)

    # Shrink log2 fold changes
    if (shrink_lfc) {
      res_shrunk <- tryCatch({
        lfcShrink(
          dds,
          contrast = contrast,
          res = res,
          type = "ashr",
          quiet = TRUE
        )
      }, error = function(e) {
        cat(sprintf("    Warning: LFC shrinkage failed: %s\n", e$message))
        res
      })

      if (!is.null(res_shrunk)) res <- res_shrunk
    }

    # Convert to dataframe
    res_df <- as.data.frame(res)
    res_df$gene <- rownames(res_df)
    res_df <- res_df[, c("gene", "baseMean", "log2FoldChange", "lfcSE", "stat", "pvalue", "padj")]

    de_results[[celltype]] <- res_df

    n_sig <- sum(res_df$padj < 0.05, na.rm = TRUE)
    cat(sprintf("    Significant genes (padj<0.05): %d\n", n_sig))
  }

  return(de_results)
}


#' Export DE results to CSV files
#'
#' @param de_results List of DESeq2 results from run_pseudobulk_deseq2()
#' @param output_dir Output directory
#' @param padj_threshold Adjusted p-value threshold (default: 0.05)
#' @param log2fc_threshold Absolute log2FC threshold (default: 0)
#'
#' @export
export_de_results <- function(
  de_results,
  output_dir = "results/pseudobulk_de",
  padj_threshold = 0.05,
  log2fc_threshold = 0
) {
  dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)

  cat(sprintf("\nExporting DE results to %s\n", output_dir))

  for (celltype in names(de_results)) {
    results_df <- de_results[[celltype]]

    # Export full results
    results_file <- file.path(output_dir, sprintf("%s_deseq2_results.csv", celltype))
    write.csv(results_df, results_file, row.names = FALSE)

    # Export significant genes
    sig_df <- results_df %>%
      filter(padj < padj_threshold, abs(log2FoldChange) > log2fc_threshold) %>%
      arrange(padj)

    sig_file <- file.path(output_dir, sprintf("%s_deseq2_sig.csv", celltype))
    write.csv(sig_df, sig_file, row.names = FALSE)

    cat(sprintf("  %s: %d significant genes\n", celltype, nrow(sig_df)))
  }
}


#' Create volcano plot for DE results
#'
#' @param results_df DESeq2 results dataframe
#' @param celltype Cell type name
#' @param output_dir Output directory
#' @param padj_threshold Adjusted p-value threshold (default: 0.05)
#' @param log2fc_threshold Log2FC threshold (default: 0.5)
#'
#' @export
plot_volcano <- function(
  results_df,
  celltype,
  output_dir = "results/pseudobulk_de",
  padj_threshold = 0.05,
  log2fc_threshold = 0.5
) {
  dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)

  # Prepare data
  plot_df <- results_df %>%
    mutate(
      neg_log10_padj = -log10(padj),
      significance = case_when(
        padj < padj_threshold & abs(log2FoldChange) > log2fc_threshold ~ "Significant",
        TRUE ~ "NS"
      )
    ) %>%
    filter(!is.na(padj))

  # Create plot
  p <- ggplot(plot_df, aes(x = log2FoldChange, y = neg_log10_padj)) +
    geom_point(aes(color = significance), alpha = 0.5, size = 1) +
    geom_hline(yintercept = -log10(padj_threshold), linetype = "dashed", color = "red") +
    geom_vline(xintercept = c(-log2fc_threshold, log2fc_threshold),
               linetype = "dashed", color = "red") +
    scale_color_manual(values = c("NS" = "#CCCCCC", "Significant" = "#E31A1C")) +
    labs(
      title = sprintf("Volcano Plot: %s", celltype),
      x = "log2 Fold Change",
      y = "-log10(adjusted p-value)"
    ) +
    theme_prism()

  # Save
  ggsave(
    file.path(output_dir, sprintf("%s_volcano.svg", celltype)),
    plot = p,
    width = 8,
    height = 6,
    dpi = 300
  )
}


#' Create MA plot for DE results
#'
#' @param results_df DESeq2 results dataframe
#' @param celltype Cell type name
#' @param output_dir Output directory
#' @param padj_threshold Adjusted p-value threshold (default: 0.05)
#'
#' @export
plot_ma <- function(
  results_df,
  celltype,
  output_dir = "results/pseudobulk_de",
  padj_threshold = 0.05
) {
  dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)

  # Prepare data
  plot_df <- results_df %>%
    mutate(
      significance = ifelse(padj < padj_threshold, "Significant", "NS")
    ) %>%
    filter(!is.na(padj), baseMean > 0)

  # Create plot
  p <- ggplot(plot_df, aes(x = log10(baseMean), y = log2FoldChange)) +
    geom_point(aes(color = significance), alpha = 0.5, size = 1) +
    geom_hline(yintercept = 0, linetype = "dashed", color = "black") +
    scale_color_manual(values = c("NS" = "#CCCCCC", "Significant" = "#E31A1C")) +
    labs(
      title = sprintf("MA Plot: %s", celltype),
      x = "Mean Expression (log10)",
      y = "log2 Fold Change"
    ) +
    theme_prism()

  # Save
  ggsave(
    file.path(output_dir, sprintf("%s_ma.svg", celltype)),
    plot = p,
    width = 8,
    height = 6,
    dpi = 300
  )
}


# Example usage
if (FALSE) {
  # Example workflow
  cat("Example pseudobulk DE workflow:\n")
  cat("1. Aggregate: pb <- aggregate_to_pseudobulk(seurat_obj, 'sample_id', 'cell_type')\n")
  cat("2. Run DESeq2: de_results <- run_pseudobulk_deseq2(pb, '~ condition', c('condition', 'treated', 'control'))\n")
  cat("3. Export: export_de_results(de_results)\n")
  cat("4. Plot: plot_volcano(de_results[['CellType']], 'CellType')\n")
}
